src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -87,47 +93,54 @@ private static async Task<List<ProjectSystemReferenceInfo>> GetAllReferencesInCo
+            CancellationToken cancellationToken)
+        {
             cancellationToken.ThrowIfCancellationRequested();
 
             var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath, cancellationToken);
 
             var referenceHandler = s_mapReferenceTypeToHandler[referenceUpdate.ReferenceInfo.ReferenceType];
 
+            IProjectSystemUpdateReferenceOperation command = s_nullCommand;
             if (referenceHandler != null)
             {
-                wasUpdated = await ApplyActionAsync(referenceUpdate, referenceHandler, activeConfiguredProject, cancellationToken);
+                command = CreateCommand(referenceUpdate, referenceHandler, activeConfiguredProject, cancellationToken);
             }
 
-            return wasUpdated;
+            return command;
         }
 
-        private static async Task<bool> ApplyActionAsync(ProjectSystemReferenceUpdate referenceUpdate, AbstractReferenceHandler referenceHandler, 
+        private IProjectSystemUpdateReferenceOperation? CreateCommand(ProjectSystemReferenceUpdate referenceUpdate, AbstractReferenceHandler referenceHandler,
             ConfiguredProject selectedConfiguredProject, CancellationToken cancellationToken)
         {
-            bool wasUpdated = false;
-
             cancellationToken.ThrowIfCancellationRequested();
 
-            if (referenceUpdate.Action == ProjectSystemUpdateAction.SetTreatAsUsed ||
-                referenceUpdate.Action == ProjectSystemUpdateAction.UnsetTreatAsUsed)
+            IProjectSystemUpdateReferenceOperation? command;
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.SetTreatAsUsed)
             {
-                wasUpdated =
-                    await referenceHandler.UpdateReferenceAsync(selectedConfiguredProject, referenceUpdate, cancellationToken);
+                command = referenceHandler.CreateSetAttributeCommand(selectedConfiguredProject, referenceUpdate);
+            }
+            else if (referenceUpdate.Action == ProjectSystemUpdateAction.UnsetTreatAsUsed)
+            {
+                command = referenceHandler.CreateUnSetAttributeCommand(selectedConfiguredProject, referenceUpdate);
             }
             else
             {
-                if (await referenceHandler.CanRemoveReferenceAsync(selectedConfiguredProject, referenceUpdate, cancellationToken))
-                {
-                    await referenceHandler.RemoveReferenceAsync(selectedConfiguredProject, referenceUpdate.ReferenceInfo);
-                    wasUpdated = true;
-                }
+                // ProjectSystemUpdateAction.Remove:
+                command = referenceHandler.CreateRemoveReferenceCommand(selectedConfiguredProject, referenceUpdate);
             }
-
-            return wasUpdated;
+            return command;
         }
```

Comment:
Similarly to my previous comment, entire method could be 1 statement:
```suggestion
        private IProjectSystemUpdateReferenceOperation? CreateCommand(ProjectSystemReferenceUpdate referenceUpdate, AbstractReferenceHandler referenceHandler,
            ConfiguredProject selectedConfiguredProject, CancellationToken cancellationToken) => referenceUpdate.Action switch
        {
            ProjectSystemUpdateAction.SetTreatAsUsed => referenceHandler.CreateSetAttributeCommand(selectedConfiguredProject, referenceUpdate),
            ProjectSystemUpdateAction.UnsetTreatAsUsed => referenceHandler.CreateUnSetAttributeCommand(selectedConfiguredProject, referenceUpdate),
            _ => referenceHandler.CreateRemoveReferenceCommand(selectedConfiguredProject, referenceUpdate)
        };
```